controllers/client_proxy.go

```diff
@@ -0,0 +1,143 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var comps []string
+	for _, item := range objList.Items {
+		name := item.GetName()
+
+		if strings.HasPrefix(name, prefix) {
+			comps = append(comps, name)
+		}
+	}
+
+	return comps, nil
+}
+
+// ListResources return only RBAC and Deployoments as this is used by:
+// - from Delete return just RBAC, we don't delete CRDs, Namespaces and resources in the namespace.
+// - from certmanager just return the resource that has a version annontation.
+func (k *controllerProxy) ListResources(labels map[string]string, namespaces ...string) ([]unstructured.Unstructured, error) {
+	resourceList := []*metav1.APIResourceList{
+		{
+			GroupVersion: "v1",
+			APIResources: []metav1.APIResource{
+				{Kind: "Secret", Namespaced: true},
+				{Kind: "ConfigMap", Namespaced: true},
+				{Kind: "Service", Namespaced: true},
+			},
+		},
+		{
+			GroupVersion: "apps/v1",
+			APIResources: []metav1.APIResource{
+				{Kind: "DaemonSet", Namespaced: true},
+				{Kind: "Deployment", Namespaced: true},
+			},
+		},
+		{
+			GroupVersion: "admissionregistration.k8s.io/v1",
+			APIResources: []metav1.APIResource{
+				{Kind: "ValidatingWebhookConfiguration", Namespaced: true},
+				{Kind: "MutatingWebhookConfiguration", Namespaced: true},
+			},
+		},
+		{
+			GroupVersion: "cert-manager.io/v1",
+			APIResources: []metav1.APIResource{
+				{Kind: "Certificate", Namespaced: true},
+				{Kind: "CertificateRequest", Namespaced: true},
+				{Kind: "Issuer", Namespaced: true},
+			},
+		},
+	}
+
+	var ret []unstructured.Unstructured
+	for _, resourceGroup := range resourceList {
+		for _, resourceKind := range resourceGroup.APIResources {
+			if resourceKind.Namespaced {
+				for _, namespace := range namespaces {
+					objList, err := listObjByGVK(k.ctrlClient, resourceGroup.GroupVersion, resourceKind.Kind, []client.ListOption{client.MatchingLabels(labels), client.InNamespace(namespace)})
+					if err != nil {
+						return nil, err
+					}
+					klog.V(3).InfoS("listed", "kind", resourceKind.Kind, "count", len(objList.Items))
+					ret = append(ret, objList.Items...)
+				}
+			} else {
+				objList, err := listObjByGVK(k.ctrlClient, resourceGroup.GroupVersion, resourceKind.Kind, []client.ListOption{client.MatchingLabels(labels)})
+				if err != nil {
+					return nil, err
+				}
+				klog.V(3).InfoS("listed", "kind", resourceKind.Kind, "count", len(objList.Items))
+				ret = append(ret, objList.Items...)
+			}
+		}
+	}
+	return ret, nil
+}
+
+func listObjByGVK(c client.Client, groupVersion, kind string, options []client.ListOption) (*unstructured.UnstructuredList, error) {
+	ctx := context.TODO()
+	objList := new(unstructured.UnstructuredList)
+	objList.SetAPIVersion(groupVersion)
+	objList.SetKind(kind)
+
+	if err := c.List(ctx, objList, options...); err != nil {
+		// not all clusters (including unit tests) will have cert-manager CRDs.
+		if _, ok := err.(*meta.NoKindMatchError); !ok {
```

Comment:
Rather than a type check, perhaps we should use `errors.Is` which will unwrap errors in the case it's been wrapped? (It's also generally nicer to read)